lib/JSON/Validator.pm

```diff
@@ -60,11 +60,18 @@ has ua => sub {
   $ua;
 };
 
+sub _this_fqn {
+  my ($self) = @_;
+  my %schemas = reverse %{ $self->{schemas} || {} };
+  $schemas{ $self->{schema}{data} || '' } // '';
+}
```

Comment:
I'm not sure if I'm a huge fan of this method. Could we rather do something like this:

```
my $root_schema = $self->schema->data;
for my $fqn (keys %{$self->{schemas}}) {
  return $fqn if $self->{schemas}{$fqn} eq $root_schema; 
}
return '';
```

Reply:
I did consider a for-loop, or perhaps a `grep`. But then I thought that would be a slightly-harder to understand, and slower, way to achieve what a hash is designed for. Do you really insist on a for-loop?

Reply:
How about this then: (Look for `root_schema_id`)

```diff
diff --git a/lib/JSON/Validator.pm b/lib/JSON/Validator.pm
index 6cc4ccd..e24ae49 100644
--- a/lib/JSON/Validator.pm
+++ b/lib/JSON/Validator.pm
@@ -66,7 +66,7 @@ sub bundle {
   my @topics = ([undef, my $bundle = {}]);
   my ($cloner, $tied);

-  $topics[0][0] = $args->{schema} ? $self->_reset->_resolve($args->{schema}) : $self->schema->data;
+  $topics[0][0] = $args->{schema} ? $self->_resolve($args->{schema}) : $self->schema->data;

   if ($args->{replace}) {
     $cloner = sub {
@@ -152,7 +152,7 @@ sub get {

 sub load_and_validate_schema {
   my ($self, $spec, $args) = @_;
-  $spec = $self->_reset->_resolve($spec);
+  $spec = $self->_resolve($spec);
   my @errors = $self->new(%$self)->schema($args->{schema} || SPECIFICATION_URL)->validate($spec);
   confess join "\n", "Invalid JSON specification $spec:", map {"- $_"} @errors if @errors;
   $self->{schema} = Mojo::JSON::Pointer->new($spec);
@@ -162,7 +162,7 @@ sub load_and_validate_schema {
 sub schema {
   my $self = shift;
   return $self->{schema} unless @_;
-  $self->{schema} = Mojo::JSON::Pointer->new($self->_reset->_resolve(shift));
+  $self->{schema} = Mojo::JSON::Pointer->new($self->_resolve(shift));
   return $self;
 }

@@ -328,18 +328,15 @@ sub _report_schema {
   push @{$self->{report}}, [(('  ') x $self->{grouped}) . ('<<<'), $path || '/', $type, D $schema];
 }

-sub _reset {
-  delete $_[0]->{schemas}{''};
-  $_[0]->{level} = 0;
-  $_[0];
-}
-
 # _resolve() method is used to convert all "id" into absolute URLs and
 # resolve all the $ref's that we find inside JSON Schema specification.
 sub _resolve {
   my ($self, $schema) = @_;
   my ($id, $resolved, @refs);

+  local $self->{level} = $self->{level} || 0;
+  delete $_[0]->{schemas}{''} unless $self->{level};
+
   if (ref $schema eq 'HASH') {
     $id = $schema->{id} // '';
     return $resolved if $resolved = $self->{schemas}{$id};
@@ -352,11 +349,13 @@ sub _resolve {
     $id = $schema->{id} if $schema->{id};
   }

-  if (!$self->{level}++ and my $id = $schema->{id}) {
-    confess "Root schema cannot have a fragment in the 'id'. ($id)" if $id =~ /\#./;
-    confess "Root schema cannot have a relative 'id'. ($id)" if $id and $id !~ /^\w+:/;
+  unless ($self->{level}) {
+    my $rid = $self->{root_schema_id} = $schema->{id} // '';
+    confess "Root schema cannot have a fragment in the 'id'. ($rid)" if $rid and $rid =~ /\#./;
+    confess "Root schema cannot have a relative 'id'. ($rid)" if $rid and $rid !~ /^\w+:/;
   }

+  $self->{level}++;
   $self->_register_schema($schema, $id);

   my @topics = ([$schema, Mojo::URL->new($id)]);
```

I want to apply this patch either way, to avoid the whole `_reset()` method, but you're welcome to use this in this PR.

Reply:
The above also needed this, which I've applied on the branch of this PR together with your patch:

```diff
diff --git a/lib/JSON/Validator.pm b/lib/JSON/Validator.pm
index daf30f1..ace94d3 100644
--- a/lib/JSON/Validator.pm
+++ b/lib/JSON/Validator.pm
@@ -351,8 +351,11 @@ sub _resolve {
 
   unless ($self->{level}) {
     my $rid = $self->{root_schema_url} = $schema->{id} // $id;
-    confess "Root schema cannot have a fragment in the 'id'. ($rid)" if $rid and $rid =~ /\#./;
-    confess "Root schema cannot have a relative 'id'. ($rid)" if $rid and $rid !~ /^\w+:/;
+    if ($rid) {
+      confess "Root schema cannot have a fragment in the 'id'. ($rid)" if $rid =~ /\#./;
+      confess "Root schema cannot have a relative 'id'. ($rid)"
+        unless $rid =~ /^\w+:/ or -e $rid or $rid =~ m!^/!;
+    }
   }
 
   $self->{level}++;
```